compiler/qsc_qasm/src/semantic/types.rs

```diff
@@ -263,25 +340,163 @@ impl Type {
     }
 }
 
+/// This function builds the indexed type of a given type.
+///
+/// Its first argument is a function that builds the base type of an array,
+/// which is only used if the array is fully indexed, and the result of the
+/// indexing operation is a scalar type.
+///
+/// Its second argument is a function that builds the array type given the
+/// new dims of the array after indexing, and the base type of the array.
+///
+/// Finally it takes the dimensions of the array being indexed and the indices
+/// used to index it. The function returns the type of the result after indexing.
 fn indexed_type_builder(
-    ty: impl Fn() -> Type,
-    ty_array: impl Fn(ArrayDimensions) -> Type,
+    base_ty_builder: impl Fn() -> Type,
+    array_ty_builder: impl Fn(ArrayDimensions) -> Type,
     dims: &ArrayDimensions,
+    indices: &[super::ast::Index],
 ) -> Type {
-    match dims.clone() {
-        ArrayDimensions::One(_) => ty(),
-        ArrayDimensions::Two(d1, _) => ty_array(ArrayDimensions::One(d1)),
-        ArrayDimensions::Three(d1, d2, _) => ty_array(ArrayDimensions::Two(d1, d2)),
-        ArrayDimensions::Four(d1, d2, d3, _) => ty_array(ArrayDimensions::Three(d1, d2, d3)),
-        ArrayDimensions::Five(d1, d2, d3, d4, _) => ty_array(ArrayDimensions::Four(d1, d2, d3, d4)),
-        ArrayDimensions::Six(d1, d2, d3, d4, d5, _) => {
-            ty_array(ArrayDimensions::Five(d1, d2, d3, d4, d5))
+    if matches!(dims, ArrayDimensions::Err) {
+        return Type::Err;
+    }
+
+    // By this point it's guaranteed that the number of indices is
+    // less or equal to the number of dims.
+    assert!(dims.num_dims() >= indices.len());
+
+    let mut indexed_dims = Vec::new();
+
+    for (dim, index) in dims.clone().into_iter().clone().zip(indices) {
+        match index {
+            Index::Expr(..) => (),
+
+            // If we have a range we need to compute the size of the slice
+            Index::Range(range) => {
+                if let Some(slice_size) = compute_slice_size(range, dim) {
+                    indexed_dims.push(slice_size);
+                } else {
+                    return Type::Err;
+                }
+            }
         }
-        ArrayDimensions::Seven(d1, d2, d3, d4, d5, d6, _) => {
-            ty_array(ArrayDimensions::Six(d1, d2, d3, d4, d5, d6))
+    }
+
+    // These are the remaining dimensions after applying all indices.
+    let not_indexed_dims = dims
+        .clone()
+        .into_iter()
+        .rev()
+        .take(dims.num_dims() - indices.len())
+        .rev();
```

Comment:
I think this can be simplified into a skip:

```suggestion
        .skip(indices.len());
```